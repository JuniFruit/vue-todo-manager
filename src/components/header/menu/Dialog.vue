<template>
    <v-dialog width="600px" v-model="isOpen">
        <template v-slot:activator="{ on, attrs }">
            <v-btn color="light-blue accent-3 text-button depressed " dark v-bind="attrs" v-on="on">
                Create a task
            </v-btn>
        </template>

        <v-card>
            <v-card-title class="text-h5 white--text  primary ">
                Add a new project
            </v-card-title>

            <v-card-text class="mt-5">
                <v-form class="px-3" @submit.prevent="submit" ref="form">
                    <v-text-field name="Title" label="Title" v-model="title" :rules="inputRules"
                        prepend-icon="mdi-account"></v-text-field>
                    <v-textarea name="Information" label="Info" v-model="info" :rules="inputRules"
                        prepend-icon="mdi-pencil"></v-textarea>
                    <v-menu offset-y min-width="auto">
                        <template v-slot:activator="{ on, attrs }">
                            <v-text-field v-on="on" v-bind="attrs" :rules="inputRules" name="Due" :value="fomatted"
                                prepend-icon="mdi-calendar" label="Due date"></v-text-field>
                        </template>
                        <v-date-picker v-model="due" color="primary lighten-2"></v-date-picker>
                    </v-menu>
                    <v-flex row justify-center>
                        <v-btn :loading="isLoading" type="submit" depressed color="primary my-5">Add a project</v-btn>

                    </v-flex>

                </v-form>
            </v-card-text>


        </v-card>
    </v-dialog>
</template>

<script lang="ts">
//@ts-nocheck
import Vue from 'vue';
import { IProjectItem } from '@/components/projects-list/List.interface';
import { Firestore } from '@/services/firestore.service';
import { mutations } from '@/store/store';
import { format } from 'date-fns';
export default Vue.extend({
    name: 'Dialog',
    data() {
        return {
            title: '',
            info: '',
            due: '',
            inputRules: [
                (v: any) => v?.length >= 3 || 'Minimum length is 3 characters'
            ],
            isLoading: false,
            isOpen: false,
            error: ''
        }
    },
    computed: {
        fomatted() {
            if (!this.due) return;
            const [y, m, d] = this.due.split('-').filter(item => item !== '-')
            return this.due ? format(new Date(+y, +m - 1, +d), "do MMM yyyy") : ''
        }
    },
    methods: {
        async submit() {
            //@ts-ignore
            if (this.$refs.form!.validate()) {
                this.isLoading = true;
                this.error = '';
                const [y, m, d] = this.due.split('-').filter(item => item !== '-')
                const project: IProjectItem = {
                    title: this.title,
                    person: 'User',
                    due: format(new Date(+y, +m - 1, +d), "do MMM yyyy"),
                    content: this.info,
                    status: 'ongoing',
                    id: '0'
                }
                const { isError, message } = await Firestore.addProject(project)
                if (isError) {
                    mutations.setToast(message, 'error')
                } else {
                    mutations.setToast('Successfully added!');
                    this.isOpen = false
                }
                this.isLoading = false

            }
        }
    }

})

</script>