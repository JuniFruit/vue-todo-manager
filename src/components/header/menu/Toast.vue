<template>
    <v-snackbar v-model="toastOpen" :color="toastType" top rounded="pill" elevation-14>
        {{ msg }}

        <template v-slot:action="{ attrs }">
            <v-btn :timeout="4000" :color="toastType" dark depressed v-bind="attrs" @click="clear">
                Close
            </v-btn>
        </template>
    </v-snackbar>
</template>


<script lang="ts">
import Vue from 'vue';
import { mutations, store } from '@/store/store';
export default Vue.extend({
    name: 'Toast',
    data() {
        return {
        }
    },
    computed: {
        toastOpen: {
            get: () => !!store.toastMsg,
            set: () => mutations.clearToast()
        },
        toastType: () => store.toastType,
        msg: () => store.toastMsg
    },
    methods: {
        clear: () => {
            mutations.clearToast()
        }
    }
})
</script>