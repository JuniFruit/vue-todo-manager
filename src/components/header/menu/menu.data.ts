interface IMenuItem {
  link: string;
  title: string;
  icon: string;
}

export const menuLinks: IMenuItem[] = [
  {
    link: "/",
    title: "Dashboard",
    icon: "mdi-view-dashboard",
  },
  {
    link: "/projects",
    title: "Projects",
    icon: "mdi-folder",
  },
  {
    link: "/team_page",
    title: "Team",
    icon: "mdi-account",
  },
];
