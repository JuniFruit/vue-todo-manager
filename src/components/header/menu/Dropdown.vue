<template>
    <v-menu offset-y>
        <template v-slot:activator="{ on, attrs }">
            <v-btn color="white" class="primary--text mx-4" v-click-outside="onClickOutside" @click="isOpen = true"
                depressed v-bind="attrs" v-on="on">
                <v-icon left>{{ isOpen ? 'mdi-chevron-up' : 'mdi-chevron-down' }}</v-icon>
                <span>Menu</span>
            </v-btn>
        </template>
        <v-list>
            <v-list-item v-for="item in menuLinks" :key="item.link" :ripple="false"
                active-class="primary lighten-4 dark--text" router :to="item.link">
                <v-list-item-title>{{ item.title }}</v-list-item-title>
            </v-list-item>
        </v-list>
    </v-menu>
</template>

<script lang="ts">
import Vue from 'vue'
import { menuLinks } from './menu.data';

export default Vue.extend({
    name: 'Dropdown',
    data() {
        return {
            menuLinks,
            isOpen: false
        }
    },
    methods: {
        onClickOutside() {
            this.isOpen = false
        }
    }
})
</script>