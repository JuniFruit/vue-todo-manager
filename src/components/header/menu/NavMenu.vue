<template>
    <div>
        <v-layout column align-center>
            <v-flex class="mt-5">
                <v-avatar size="100">
                    <img src="https://www.gravatar.com/avatar/4d2393d211ba4cd39cb26d02d1812fbd?s=140&d=retro" alt="">
                </v-avatar>
            </v-flex>
            <span class="white--text">User</span>
        </v-layout>
        <v-list>
            <v-list-item v-for="menuItem in links" :key="menuItem.link" router :to="menuItem.link">
                <v-list-item-action>
                    <v-icon class="white--text">{{ menuItem.icon }}</v-icon>
                </v-list-item-action>
                <v-list-item-content>
                    <v-list-item-title class="white--text">
                        {{ menuItem.title }}
                    </v-list-item-title>
                </v-list-item-content>
            </v-list-item>
        </v-list>
    </div>
</template>


<script lang="ts">
import { menuLinks } from './menu.data';

export default {
    data() {
        return {
            links: menuLinks
        }
    }
}


</script>