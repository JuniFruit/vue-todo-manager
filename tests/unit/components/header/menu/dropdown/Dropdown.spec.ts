import Dropdown from "@/components/header/menu/Dropdown.vue";
import { menuLinks } from "@/components/header/menu/menu.data";
import { Wrapper } from "@vue/test-utils";
import { createCustomWrapper } from "../../../../utils/general";

describe("Dropdown.vue component", () => {
  let wrapper: Wrapper<Dropdown>;

  beforeEach(() => {
    wrapper = createCustomWrapper<Dropdown>(Dropdown, {});
  });

  afterEach(() => {
    wrapper.vm.$destroy();
  });

  it("renders chevron down when not clicked", () => {
    const chevronDown = wrapper.find(".mdi-chevron-down");
    expect(chevronDown.exists()).toBe(true);
  });
  it("renders chevron up when clicked", async () => {
    const dropdownButton = wrapper.find(".v-btn");
    await dropdownButton.trigger("click");
    const chevronUp = wrapper.find(".mdi-chevron-up");
    expect(chevronUp.exists()).toBe(true);
  });

  it("opens vuetify list with links when clicked", async () => {
    const dropdownButton = wrapper.find(".v-btn");
    await dropdownButton.trigger("click");

    const list = wrapper.find(".v-list");
    expect(list).toMatchSnapshot();
    menuLinks.forEach(link => {
      expect(wrapper.html()).toMatch(link.title);
    });
  });

  it("redirects to a page when link is clicked", async () => {
    const dropdownButton = wrapper.find(".v-btn");
    await dropdownButton.trigger("click");
    const routerSpy = jest.spyOn(wrapper.vm.$router, "push");
    const link = wrapper.find("a");
    await link.trigger("click");

    expect(routerSpy).toHaveBeenCalled();
  });
});
