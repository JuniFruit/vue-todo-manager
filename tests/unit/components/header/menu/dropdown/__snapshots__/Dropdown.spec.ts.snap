// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown.vue component opens vuetify list with links when clicked 1`] = `
<div class="v-list v-sheet theme--light"><a href="/" aria-current="page" class="primary lighten-4 dark--text v-list-item--active v-list-item v-list-item--link theme--light" tabindex="0" router="" role="menuitem" id="list-item-27">
    <div class="v-list-item__title">Dashboard</div>
  </a><a href="/projects" class="v-list-item v-list-item--link theme--light" tabindex="0" router="" role="menuitem" id="list-item-29">
    <div class="v-list-item__title">Projects</div>
  </a><a href="/team_page" class="v-list-item v-list-item--link theme--light" tabindex="0" router="" role="menuitem" id="list-item-31">
    <div class="v-list-item__title">Team</div>
  </a></div>
`;
